tango_ros_common/tango_ros_native/src/tango_ros_node.cpp

```diff
@@ -610,26 +623,87 @@ void TangoRosNode::OnPoseAvailable(const TangoPoseData* pose) {
+          toTransformStamped(*pose, time_offset_, &area_description_T_device_);
+          area_description_T_device_.header.frame_id =
               toFrameId(TANGO_COORDINATE_FRAME_AREA_DESCRIPTION);
-          area_description_T_start_of_service_.child_frame_id =
+          area_description_T_device_.child_frame_id =
+              toFrameId(TANGO_COORDINATE_FRAME_DEVICE);
+
+          start_of_service_T_area_description_.header.stamp = ros::Time::now();
+
+          if (localization_mode_ != LocalizationMode::ODOMETRY) {
+            TangoCoordinateFramePair pair;
+            pair.base = TANGO_COORDINATE_FRAME_AREA_DESCRIPTION;
+            pair.target = TANGO_COORDINATE_FRAME_DEVICE;
+            TangoPoseData area_description_T_device;
+            TangoService_getPoseAtTime(0.0, pair, &area_description_T_device);
+            first_localized_ = (area_description_T_device.status_code == TANGO_POSE_VALID);
+            if (first_localized_) {
+              LOG(INFO) << "We are localized!";
+            }
+          }
+        } else if (localization_lost_) {
+          LOG(WARNING) << "Recovering tracking...";
+          // We lost localization, use old start_of_service_T_area_description
+          // to compute area_description_T_device.
+          geometry_msgs::TransformStamped star_of_service_T_device;
+          toTransformStamped(*pose, time_offset_, &star_of_service_T_device);
+          star_of_service_T_device.header.frame_id =
+              toFrameId(TANGO_COORDINATE_FRAME_START_OF_SERVICE);
+          star_of_service_T_device.child_frame_id =
+              toFrameId(TANGO_COORDINATE_FRAME_DEVICE);
+
+          start_of_service_T_area_description_.header.stamp = ros::Time::now();
+
+          tf::StampedTransform star_of_service_T_device_tf;
+          tf::StampedTransform start_of_service_T_area_description_tf;
+          tf::transformStampedMsgToTF(star_of_service_T_device,
+                                      star_of_service_T_device_tf);
+          tf::transformStampedMsgToTF(start_of_service_T_area_description_,
+                                      start_of_service_T_area_description_tf);
+          tf::Transform area_description_T_device_tf =
+              start_of_service_T_area_description_tf.inverse() * star_of_service_T_device_tf;
+          tf::transformStampedTFToMsg(tf::StampedTransform(
+              area_description_T_device_tf, star_of_service_T_device.header.stamp,
+              toFrameId(TANGO_COORDINATE_FRAME_AREA_DESCRIPTION),
+              toFrameId(TANGO_COORDINATE_FRAME_DEVICE)), area_description_T_device_);
+        }
+        pose_available_.notify_all();
+        pose_available_mutex_.unlock();
+      }
+    } else if (pose->frame.base == TANGO_COORDINATE_FRAME_AREA_DESCRIPTION
+        && pose->frame.target == TANGO_COORDINATE_FRAME_DEVICE
+        && first_localized_) {
+      if (pose->status_code == TANGO_POSE_VALID && pose_available_mutex_.try_lock()) {
```

Comment:
I like the first one :+1: 
The second one seems "dangerous", I prefer to keep the mutex lock explicit.